In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.preprocessing import MinMaxScaler
import tensorflow as tf
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset

Q1/Load the dataset, airfoil_self_noise.DAT, into memory

In [2]:
df = pd.read_csv('airfoil_self_noise.DAT', sep='\t', header=None)
print(df.head())

      0    1       2     3         4        5
0   800  0.0  0.3048  71.3  0.002663  126.201
1  1000  0.0  0.3048  71.3  0.002663  125.201
2  1250  0.0  0.3048  71.3  0.002663  125.951
3  1600  0.0  0.3048  71.3  0.002663  127.591
4  2000  0.0  0.3048  71.3  0.002663  127.461


Q2/Clean the data and check missing values for this dataset.

In [3]:

missing_values = df.isnull().sum()
print(missing_values)

0    0
1    0
2    0
3    0
4    0
5    0
dtype: int64


Q3/Split the data into 80% of training and 20% of the test dataset.

In [4]:
X = df.iloc[:, :-1]  
y = df.iloc[:, -1]   
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("Training dataset shape:", X_train.shape, y_train.shape)
print("Test dataset shape :", X_test.shape, y_test.shape)


Training dataset shape: (1202, 5) (1202,)
Test dataset shape : (301, 5) (301,)


 Build a simple linear regression to forecast "Scaled sound pressure level" using all other features and scikit-learn package.

In [5]:
model = LinearRegression()
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print("Mean Squared Error:", mse)
print("R-squared Score:", r2)


Mean Squared Error: 22.128643318247306
R-squared Score: 0.5582979754897279


D:\aldul\Lib\site-packages\sklearn\utils\validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
D:\aldul\Lib\site-packages\sklearn\utils\

 What's the test error for this scikit-learn regression model?
 the test error for this scikit-learn regression model is 22.128643318247306

Q4/Preprocess the data using the normalization method to convert all features into the range of [0,1].


In [6]:
scaler = MinMaxScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)


D:\aldul\Lib\site-packages\sklearn\utils\validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:605: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:614: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
D:\aldul\Lib\site-packages\sklearn\utils\validation.py:767: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse)

Q5/Build a deep learning regression model to forecast "Scaled sound pressure level" using all other features and TensorFlow. Please use only two layers of the neuron network.
You choose the number of neurons to use in the first layer

In [7]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)),
    tf.keras.layers.Dense(1) 
])
model.compile(optimizer='adam', loss='mean_squared_error')
model.fit(X_train_normalized, y_train, epochs=100, batch_size=32, validation_split=0.2)
test_loss = model.evaluate(X_test_normalized, y_test)
print("Test Loss:", test_loss)

D:\aldul\Lib\site-packages\keras\src\layers\core\dense.py:85: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - loss: 15595.2451 - val_loss: 15518.6904
Epoch 2/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 15417.8809 - val_loss: 15330.4395
Epoch 3/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 15282.2158 - val_loss: 15028.8047
Epoch 4/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 14901.5391 - val_loss: 14615.1152
Epoch 5/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 14377.7861 - val_loss: 14100.9980
Epoch 6/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 13990.1465 - val_loss: 13493.5938
Epoch 7/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 13228.6484 - val_loss: 12803.1562
Epoch 8/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 12672.0771 - val_loss: 12039.3633
Epoch 9/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 11826.9746 - val_loss: 11220.6133
Epoch 10/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 10896.6523 - val_loss: 10347.6660
Epoch 11/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 9995.5664

31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 553.8740 - val_loss: 569.7731
Epoch 67/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 570.9138 - val_loss: 557.0753
Epoch 68/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 544.3941 - val_loss: 546.3280
Epoch 69/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 547.4971 - val_loss: 534.4062
Epoch 70/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 551.0588 - val_loss: 522.6865
Epoch 71/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 517.2660 - val_loss: 510.8399
Epoch 72/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 527.0154 - val_loss: 497.8427
Epoch 73/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 487.2749 - val_loss: 485.9343
Epoch 74/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 511.0184 - val_loss: 474.6345
Epoch 75/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 464.8322 - val_loss: 463.1515
Epoch 76/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 478.6910 - val_loss: 450.8939
Epoch 77/100
31/31 ━━━━

 What's the test error for this model? is approximately 241.416748046875

Q6/Can you improve the model performance of question 5 by adjusting the number of neurons or the optimization algorithm?

In [8]:
model = tf.keras.Sequential([
    tf.keras.layers.Dense(128, activation='relu', input_shape=(X_train.shape[1],)),  
    tf.keras.layers.Dense(1) 
])
model.compile(optimizer='adamax', loss='mean_squared_error')
history = model.fit(X_train_normalized, y_train, epochs=100, batch_size=32, validation_split=0.2)
test_loss = model.evaluate(X_test_normalized, y_test)
print("Test Loss:", test_loss)


Epoch 1/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 1s 6ms/step - loss: 15587.4043 - val_loss: 15547.8691
Epoch 2/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 15462.7363 - val_loss: 15411.0000
Epoch 3/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 15355.5908 - val_loss: 15234.9375
Epoch 4/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 15141.1699 - val_loss: 15022.6240
Epoch 5/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 14971.4717 - val_loss: 14778.2031
Epoch 6/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 14644.8564 - val_loss: 14496.9346
Epoch 7/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 14348.1318 - val_loss: 14173.3613
Epoch 8/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 14033.4902 - val_loss: 13837.2314
Epoch 9/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 13634.9629 - val_loss: 13443.4258
Epoch 10/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 13312.9170 - val_loss: 12999.5996
Epoch 11/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 12860.712

31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 515.6248 - val_loss: 532.3119
Epoch 67/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 531.6867 - val_loss: 528.4968
Epoch 68/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 510.6830 - val_loss: 524.9112
Epoch 69/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 526.8582 - val_loss: 520.9191
Epoch 70/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 515.3533 - val_loss: 516.9612
Epoch 71/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 556.2944 - val_loss: 512.8697
Epoch 72/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 507.3092 - val_loss: 508.2113
Epoch 73/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 517.1626 - val_loss: 503.3065
Epoch 74/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 551.3911 - val_loss: 499.2267
Epoch 75/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 482.9218 - val_loss: 494.8491
Epoch 76/100
31/31 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 476.9716 - val_loss: 490.1405
Epoch 77/100
31/31 ━━━━

Q7/Build a deep learning regression model to forecast "Scaled sound pressure level" using all other features and PyTorch. Please use only two layers of the neuron network. You choose the number of neurons to use in the first layer. 

In [9]:

num_neurons = 128
X_train_normalized_tensor = torch.tensor(X_train_normalized, dtype=torch.float32)
X_test_normalized_tensor = torch.tensor(X_test_normalized, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train.values, dtype=torch.float32).view(-1, 1)  
y_test_tensor = torch.tensor(y_test.values, dtype=torch.float32).view(-1, 1)  

class RegressionModel(nn.Module):
    def __init__(self, input_size, num_neurons):
        super(RegressionModel, self).__init__()
        self.fc1 = nn.Linear(input_size, num_neurons)
        self.fc2 = nn.Linear(num_neurons, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return x
model = RegressionModel(input_size=X_train_normalized_tensor.shape[1], num_neurons=num_neurons)
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters())
batch_size = 32
train_dataset = TensorDataset(X_train_normalized_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_dataset = TensorDataset(X_test_normalized_tensor, y_test_tensor)
test_loader = DataLoader(test_dataset, batch_size=batch_size)
num_epochs = 20
for epoch in range(num_epochs):
    model.train()
    running_loss = 0.0
    for inputs, targets in train_loader:
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, targets)
        loss.backward()
        optimizer.step()
        running_loss += loss.item() * inputs.size(0)
    epoch_loss = running_loss / len(train_loader.dataset)
    print(f"Epoch {epoch+1}/{num_epochs}, Loss: {epoch_loss:.4f}")
model.eval()
test_loss = 0.0
with torch.no_grad():
    for inputs, targets in test_loader:
        outputs = model(inputs)
        loss = criterion(outputs, targets)
        test_loss += loss.item() * inputs.size(0)
test_loss /= len(test_loader.dataset)
print(f"Test Loss: {test_loss:.4f}")


Epoch 1/20, Loss: 15517.9404
Epoch 2/20, Loss: 15185.3444
Epoch 3/20, Loss: 14622.5822
Epoch 4/20, Loss: 13708.5254
Epoch 5/20, Loss: 12428.7822
Epoch 6/20, Loss: 10853.0105
Epoch 7/20, Loss: 9093.6079
Epoch 8/20, Loss: 7300.4589
Epoch 9/20, Loss: 5611.1962
Epoch 10/20, Loss: 4147.7630
Epoch 11/20, Loss: 2984.0078
Epoch 12/20, Loss: 2143.2135
Epoch 13/20, Loss: 1588.1026
Epoch 14/20, Loss: 1254.8626
Epoch 15/20, Loss: 1072.2051
Epoch 16/20, Loss: 978.3416
Epoch 17/20, Loss: 927.2351
Epoch 18/20, Loss: 895.2332
Epoch 19/20, Loss: 870.6987
Epoch 20/20, Loss: 847.6873
Test Loss: 911.9798


What's the test error for this model?is approximately 911.9798

Q8/Can you improve the model performance of question 7 by adjusting the number of neurons or the optimization algorithm?

In [10]:

def train_and_evaluate_model(input_size, num_neurons, optimizer_name, epochs):
    model = RegressionModel(input_size, num_neurons)
    
    if optimizer_name == 'adam':
        optimizer = optim.Adam(model.parameters(), lr=0.001)
    elif optimizer_name == 'adamax':
        optimizer = optim.Adamax(model.parameters(), lr=0.001)
    elif optimizer_name == 'sgd':
        optimizer = optim.SGD(model.parameters(), lr=0.001)
    elif optimizer_name == 'rmsprop':
        optimizer = optim.RMSprop(model.parameters(), lr=0.001)
    else:
        raise ValueError("Invalid optimizer name")

    train_dataset = TensorDataset(X_train_normalized_tensor, y_train_tensor)
    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

    for epoch in range(epochs):
        model.train()
        running_loss = 0.0
        for inputs, targets in train_loader:
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
            running_loss += loss.item() * inputs.size(0)
        epoch_loss = running_loss / len(train_loader.dataset)
        print(f"Epoch {epoch+1}/{epochs}, Loss: {epoch_loss:.4f}")

    model.eval()
    test_loss = 0.0
    with torch.no_grad():
        for inputs, targets in test_loader:
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            test_loss += loss.item() * inputs.size(0)
    test_loss /= len(test_loader.dataset)
    print(f"Test Loss: {test_loss:.4f}")
    return test_loss


num_neurons_list = [16, 32, 64]
optimizers = ['adam', 'adamax', 'sgd', 'rmsprop']
epochs = 20

best_num_neurons = None
best_optimizer = None
best_test_error = float('inf')

for num_neurons in num_neurons_list:
    for optimizer_name in optimizers:
        try:
            test_error = train_and_evaluate_model(X_train_normalized_tensor.shape[1], num_neurons, optimizer_name, epochs)
            print(f"Test error for {num_neurons} neurons and {optimizer_name} optimizer:", test_error)
        
            
            if test_error < best_test_error:
                best_test_error = test_error
                best_num_neurons = num_neurons
                best_optimizer = optimizer_name
        except ValueError as e:
            print(e)
            continue

print("Best hyperparameters:")
print("Number of neurons in the first layer:", best_num_neurons)
print("Optimizer:", best_optimizer)
print("Best test error:", best_test_error)


Epoch 1/20, Loss: 15563.2406
Epoch 2/20, Loss: 15504.5122
Epoch 3/20, Loss: 15423.2436
Epoch 4/20, Loss: 15306.9082
Epoch 5/20, Loss: 15151.5529
Epoch 6/20, Loss: 14960.4816
Epoch 7/20, Loss: 14735.3073
Epoch 8/20, Loss: 14477.6318
Epoch 9/20, Loss: 14189.2355
Epoch 10/20, Loss: 13871.9868
Epoch 11/20, Loss: 13528.0554
Epoch 12/20, Loss: 13158.9913
Epoch 13/20, Loss: 12768.7297
Epoch 14/20, Loss: 12358.7037
Epoch 15/20, Loss: 11931.6346
Epoch 16/20, Loss: 11490.2406
Epoch 17/20, Loss: 11036.8328
Epoch 18/20, Loss: 10574.5693
Epoch 19/20, Loss: 10105.7506
Epoch 20/20, Loss: 9633.0439
Test Loss: 9486.3830
Test error for 16 neurons and adam optimizer: 9486.383049340739
Epoch 1/20, Loss: 15610.5673
Epoch 2/20, Loss: 15563.7920
Epoch 3/20, Loss: 15515.5507
Epoch 4/20, Loss: 15457.6945
Epoch 5/20, Loss: 15386.9189
Epoch 6/20, Loss: 15302.4840
Epoch 7/20, Loss: 15203.2070
Epoch 8/20, Loss: 15091.2490
Epoch 9/20, Loss: 14965.7941
Epoch 10/20, Loss: 14824.2347
Epoch 11/20, Loss: 14670.0905
Epoc